### ISSUE TITLE
Differentiability with respect to cosmology

### ISSUE THREAD

**Issue by EiffL** [modichirag/flowpm#46]
One of the obvious things we want to be able to do with FlowPM is to get gradients with respect to cosmology. Currently, we are relying on the [background.py](https://github.com/rainwoodman/fastpm-python/blob/master/fastpm/background.py) file from the python implementation of FastPM, so we can't backpropagate with respect to cosmological parameters.
 
The goal of this issue is to add a pure tensorflow implementation of cosmology computations that will allow us to replace all the calls to background.py. I have created a `tfcosmo` branch: https://github.com/modichirag/flowpm/tree/tfcosmo to work on this implementation.

I have made a small list of things we want to be able to compute in pure tensorflow:
  - [x] : Analytic initial matter power spectrum, probably Eisenstein & Hu
  - [x] : Linear growth function [D1](https://github.com/modichirag/flowpm/blob/04f111e270af5a05da519c88061f31e828e7081b/flowpm/background.py#L41)
  - [x] : Hubble function [E](https://github.com/modichirag/flowpm/blob/04f111e270af5a05da519c88061f31e828e7081b/flowpm/background.py#L165)

Other functions used in the nbody code can easily be derived once we have this in place.

Very concretely, we can start building a `tfbackground.py` file and add these features as we go, until we can replace the `background.py` file altogether.

For "inspiration" we can reuse code from the jax-cosmo project, where all of this is implemented, but alas in Jax and not TensorFlow :  https://github.com/DifferentiableUniverseInitiative/jax_cosmo/blob/master/jax_cosmo/background.py

I'm happy to start looking at this with @dlanzieri

**Comment by EiffL** [comment]
Thanks @dlanzieri for adding your efunction_background notebook on https://github.com/modichirag/flowpm/tree/tfcosmo  :-)
Could you move it to the notebooks folder? 
Also, could you create a `tfbackground.py`file in the flowpm/flowpm folder? Where we can add your function implementations

**Comment by dlanzieri** [comment]
@EiffL Of course, I'm going to do that!

**Comment by EiffL** [comment]
Hi @dlanzieri , just to let you know, I have added a file that contains your new functions in a `tfbackground.py` file: https://github.com/modichirag/flowpm/blob/tfcosmo/flowpm/tfbackground.py

**Comment by dlanzieri** [comment]
I added two notebooks code to the tfcosmo branch: efunction_background.ipynb and growth_function.ipynb. In the first function I built a pure Tensorflow implementation of the Hubble function E. In the second one, I built a Tensorflow implementation of the Linear growth function D1 and of the Linear growth rate D'.

**Comment by EiffL** [comment]
Thanks @dlanzieri :-) It's looking very good, next step is to try to build some tests to check your equations and implementation.
You can try to use the existing functions from `background.py`, for instance to get D1 as a function of scale factor a, and check that you get the same results.

**Comment by Maxelee** [comment]
Here are my comparisons between EH and CLASS. These are the C_l's created by MADLens using an EH and a CLASS initial power spectrum. The first plot is just the division of the two, the second is them plotted with a theoretical C_l, and 2sigma shading from cosmic variance. I forgot to add axis labels but x-axis is L, y is C_l. let me know if this helps (or does not help)!


<img width="673" alt="Screen Shot 2020-10-28 at 11 03 35 AM" src="https://user-images.githubusercontent.com/47232440/97477879-6da91980-190d-11eb-9061-e04401b1e5ea.png">
<img width="753" alt="Screen Shot 2020-10-28 at 11 04 08 AM" src="https://user-images.githubusercontent.com/47232440/97477926-7863ae80-190d-11eb-8dbc-5220910bfe1e.png">

**Comment by EiffL** [comment]
Great progress @dlanzieri !   
So now, what we want to do, is to look at `tfpm.py` and try to replace all of the calls to the old functions from `background.py` to the new functions you have in `tfbackground.py`. You will see that maybe you need to add some additional functions as well.

For instance, take a look at:
https://github.com/modichirag/flowpm/blob/3dfc6bad346e6e923f1c5d2877d46c47382e8624/flowpm/tfpm.py#L148
which computes the 1LPT evolution down to a given scale factor. 

You could try to setup a notebook to just run this LPT step, based on the  example notebook: https://github.com/modichirag/flowpm/blob/master/notebooks/flowpm_blog.ipynb
And start replacing the functions you already have implemented to see if things still work
